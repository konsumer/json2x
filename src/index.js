export const allArraySame = obj => {
  for (var i = 1; i < obj.length; i++) {
    if (obj[i] !== obj[0]) return false
  }
  return true
}

export const isFloat = n => n === +n && n !== (n | 0)

export const getBestType = obj => {
  const t = ({}).toString.call(obj).split(' ')[1].slice(0, -1).toLowerCase()
  switch (t) {
    case 'object': return Array.isArray(obj) ? 'array' : 'object'
    case 'number': return isFloat(obj) ? 'float' : 'int'
    default: return t
  }
}

const getAverageFromTypes = types => {
  if (allArraySame(types)) {
    return types[0]
  }
  // if any are a string, return string
  if (types.indexOf('string') !== -1) {
    return 'string'
  }
  // if any are a float, return float
  if (types.indexOf('float') !== -1) {
    return 'float'
  }
  return types[0]
}

export const getAverageType = values => getAverageFromTypes(values.map(getBestType))

/**
 * Get types from an Object filled with similar objects
 * @param  object data An Object filled with similar objects
 * @return object      Type info
 */
export const getTypes = data => {
  const fields = {}
  const out = {}
  const objs = {}
  const arrays = {}
  data.forEach(record => {
    Object.keys(record).forEach(k => {
      const type = getBestType(record[k])
      if (type === 'array') {
        arrays[k] = arrays[k] || []
        return arrays[k].push(getAverageType(record[k]))
      }
      if (type === 'object') {
        objs[k] = objs[k] || []
        return objs[k].push(record[k])
      }
      fields[k] = fields[k] || []
      return fields[k].push(record[k])
    })
  })

  Object.keys(fields).forEach(k => {
    out[k] = getAverageType(fields[k])
  })

  Object.keys(objs).forEach(k => {
    out[k] = getTypes(objs[k])
  })

  Object.keys(arrays).forEach(k => {
    out[k] = [getAverageFromTypes(arrays[k])]
  })

  return out
}

module.exports = { allArraySame, getBestType, getAverageType, getTypes }
