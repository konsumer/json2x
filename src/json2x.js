#!/usr/bin/env babel-node

import { promisify } from 'util'
import { basename } from 'path'
import { readFileSync } from 'fs'
import yargs from 'yargs'
import globR from 'glob-require'

require('babel-polyfill')

const noop = () => {}
const g = promisify(globR)

const main = async () => {
  const commands = await g('**/*.js', {cwd: `${__dirname}/generators`})

  commands.forEach(({path, exports: {description, options, run}}) => {
    const name = basename(path, '.js')
    yargs.command(
      `${name} <MODEL> <FILES...>`,
      description,
      options || noop,
      opt => {
        opt.raw = {}
        opt.data = {}
        opt.FILES.forEach(source => {
          opt.raw[source] = readFileSync(source).toString()
          opt.data[source] = JSON.parse(opt.raw[source])
        })
        Promise.resolve(run(opt)).then(results => console.log(results))
      }
    )
    yargs.example(`$0 ${name} -h`, `Get some help with the ${name} command`)
  })

  yargs // eslint-disable-line
    .demandCommand()
    .help()
    .version()
    .alias('h', 'help')
    .alias('v', 'version')
    .wrap(process.stdout.columns)
    .argv
}
main()
