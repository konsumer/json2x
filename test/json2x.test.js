/* global describe, it, expect */

import { getCommands, g } from '../src/json2x'
import { getAverageType, getTypes } from '../src/index'

describe('json2x', () => {
  it('should get the commands I expect', async () => {
    expect(await getCommands()).toMatchSnapshot()
  })

  it('should get average from types for [1, 2, 3.5, 5]', () => {
    expect(getAverageType([1, 2, 3.5, 5])).toMatchSnapshot()
  })

  it('should get average from types for [1, 2, 3.0, 5]', () => {
    expect(getAverageType([1, 2, 3.0, 5])).toMatchSnapshot()
  })

  it('should get average from types for ["a", 1, 2, 3.0, 5]', () => {
    expect(getAverageType(['a', 1, 2, 3.0, 5])).toMatchSnapshot()
  })

  it('should get types from a more complex objects', async () => {
    const mock = (await g('*.json', {cwd: `${__dirname}/mock`})).map(({ exports }) => exports)
    expect(getTypes(mock)).toMatchSnapshot()
  })
})
