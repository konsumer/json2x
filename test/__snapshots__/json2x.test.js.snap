// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json2x should get average from types for ["a", 1, 2, 3.0, 5] 1`] = `"string"`;

exports[`json2x should get average from types for [1, 2, 3.0, 5] 1`] = `"int"`;

exports[`json2x should get average from types for [1, 2, 3.5, 5] 1`] = `"float"`;

exports[`json2x should get the commands I expect 1`] = `
Array [
  Object {
    "exports": Object {
      "description": "Generate GraphQL types from your data",
      "makeType": [Function],
      "options": [Function],
      "run": [Function],
    },
    "path": "/Users/konsumer/Documents/otherdev/json2x/src/generators/graphql.js",
  },
  Object {
    "exports": Object {
      "description": "Generate JSONSchema from your data",
      "run": [Function],
    },
    "path": "/Users/konsumer/Documents/otherdev/json2x/src/generators/jsonschema.js",
  },
  Object {
    "exports": Object {
      "description": "Generate GraphQL types from your data",
      "makeType": [Function],
      "options": [Function],
      "run": [Function],
    },
    "path": "/Users/konsumer/Documents/otherdev/json2x/src/generators/proto.js",
  },
]
`;

exports[`json2x should get types from a more complex objects 1`] = `
Object {
  "address": Object {
    "city": "string",
    "geo": Object {
      "lat": "float",
      "lng": "float",
    },
    "street": "string",
    "suite": "string",
    "zipcode": "string",
  },
  "company": Object {
    "bs": "string",
    "catchPhrase": "string",
    "name": "string",
  },
  "doing": "string",
  "email": "string",
  "id": "string",
  "lastSaid": "string",
  "name": "string",
  "phone": "string",
  "testnumobjnums": Object {
    "deeper": Object {
      "arr": Array [
        "float",
      ],
    },
    "num1": "int",
    "num2": "int",
    "num3": "float",
  },
  "username": "string",
  "website": "string",
}
`;
